File: python/council-sdk/council/command.py
"""AGP Command operations — agent registration, governance, deployments & fleets."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from council.transport import Transport


class CommandNamespace:
    """AGP Command operations for agent registration, governance actions,
    deployments, fleets, and activity streams.

    Usage::

        agent = await client.command.register_agent(
            name="Bot",
            organization_id="org_1",
            agent_class="DIGITAL",
            digital={"model": "gpt-4", "provider": "openai", "runtime": "NODE"},
            declared_caps=["api.call"],
            governance_profile_id="gov_1",
            provenance_trust="INTERNAL",
            code_trust="AUDITED",
            network_trust="VPN",
        )
    """

    def __init__(self, transport: Transport) -> None:
        self._transport = transport

    # ── Agent Registration ──────────────────────────────────────────────

    async def register_agent(
        self,
        name: str,
        organization_id: str,
        agent_class: str,
        declared_caps: List[str],
        governance_profile_id: str,
        provenance_trust: str,
        code_trust: str,
        network_trust: str,
        description: Optional[str] = None,
        digital: Optional[Dict[str, Any]] = None,
        physical: Optional[Dict[str, Any]] = None,
        hardware_trust: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Register a new agent in the AGP registry."""
        return await self._transport.post("/agp/v1/agents/register", json={
            "name": name,
            "organizationId": organization_id,
            "agentClass": agent_class,
            "description": description,
            "digital": digital,
            "physical": physical,
            "declaredCaps": declared_caps,
            "governanceProfileId": governance_profile_id,
            "provenanceTrust": provenance_trust,
            "codeTrust": code_trust,
            "hardwareTrust": hardware_trust,
            "networkTrust": network_trust,
        })

    async def get_agent(self, agent_id: str) -> Dict[str, Any]:
        """Get an agent by ID."""
        return await self._transport.get(f"/agp/v1/agents/{agent_id}")

    async def list_agents(self, organization_id: Optional[str] = None) -> List[Dict[str, Any]]:
        """List agents, optionally filtered by organization."""
        params = f"?organizationId={organization_id}" if organization_id else ""
        return await self._transport.get(f"/agp/v1/agents{params}")

    # ── Governance Actions ──────────────────────────────────────────────

    async def request_action(
        self, agent_id: str, action: str, context: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Request a governed action on behalf of an agent."""
        return await self._transport.post("/agp/v1/governance/action", json={
            "agentId": agent_id,
            "action": action,
            "context": context,
            "protocolVersion": "agp/v1",
        })

    async def get_governance_events(self, agent_id: str) -> List[Dict[str, Any]]:
        """Get governance events for an agent."""
        return await self._transport.get(f"/agp/v1/governance/events/{agent_id}")

    # ── Deployments ─────────────────────────────────────────────────────

    async def create_deployment(
        self,
        agent_id: str,
        mode: str,
        compute_target: str,
        cloud_spec: Optional[Dict[str, Any]] = None,
        replicas: int = 1,
        schedule: Optional[str] = None,
        edge_node_id: Optional[str] = None,
        external_endpoint: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Create a new deployment for an agent."""
        return await self._transport.post("/agp/v1/deployments", json={
            "agentId": agent_id,
            "mode": mode,
            "computeTarget": compute_target,
            "cloudSpec": cloud_spec,
            "replicas": replicas,
            "schedule": schedule,
            "edgeNodeId": edge_node_id,
            "externalEndpoint": external_endpoint,
        })

    async def list_deployments(self, agent_id: str) -> List[Dict[str, Any]]:
        """List deployments for an agent."""
        return await self._transport.get(f"/agp/v1/deployments/{agent_id}")

    # ── Fleets ──────────────────────────────────────────────────────────

    async def create_fleet(
        self,
        name: str,
        organization_id: str,
        governance_profile_id: str,
        description: Optional[str] = None,
        template_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Create a new fleet."""
        return await self._transport.post("/agp/v1/fleets", json={
            "name": name,
            "organizationId": organization_id,
            "description": description,
            "templateId": template_id,
            "governanceProfileId": governance_profile_id,
        })

    async def list_fleets(self, organization_id: Optional[str] = None) -> List[Dict[str, Any]]:
        """List fleets, optionally filtered by organization."""
        params = f"?organizationId={organization_id}" if organization_id else ""
        return await self._transport.get(f"/agp/v1/fleets{params}")

    # ── Activity ────────────────────────────────────────────────────────

    async def get_activity(self, agent_id: str) -> List[Dict[str, Any]]:
        """Get the activity stream for an agent."""
        return await self._transport.get(f"/agp/v1/activity/{agent_id}")

File: typescript/src/namespaces/command.ts
import { Transport } from "../transport.js";

export interface RegisterAgentInput {
  name: string;
  organizationId: string;
  agentClass: "DIGITAL" | "PHYSICAL" | "HYBRID";
  description?: string;
  digital?: { model: string; provider: string; runtime: string };
  physical?: {
    embodimentType: string;
    sensorManifest: any[];
    actuatorManifest: any[];
  };
  declaredCaps: string[];
  governanceProfileId: string;
  provenanceTrust: string;
  codeTrust: string;
  hardwareTrust?: string;
  networkTrust: string;
}

export interface ActionRequestInput {
  agentId: string;
  action: string;
  context: Record<string, unknown>;
}

export interface CreateDeploymentInput {
  agentId: string;
  mode: "PERSISTENT" | "SCHEDULED" | "ON_DEMAND";
  computeTarget: "MANAGED" | "EXTERNAL" | "EDGE";
  cloudSpec?: Record<string, unknown>;
  replicas?: number;
  schedule?: string;
  edgeNodeId?: string;
  externalEndpoint?: string;
}

export interface CreateFleetInput {
  name: string;
  organizationId: string;
  description?: string;
  templateId?: string;
  governanceProfileId: string;
}

/**
 * AGP Command operations — agent registration, governance actions,
 * deployments, fleets, and activity streams.
 *
 * @example
 * ```typescript
 * const agent = await client.command.registerAgent({
 *   name: 'Bot', organizationId: 'org_1', agentClass: 'DIGITAL',
 *   digital: { model: 'gpt-4', provider: 'openai', runtime: 'NODE' },
 *   declaredCaps: ['api.call'], governanceProfileId: 'gov_1',
 *   provenanceTrust: 'INTERNAL', codeTrust: 'AUDITED', networkTrust: 'VPN',
 * });
 * ```
 */
export class CommandNamespace {
  constructor(private transport: Transport) {}

  // ── Agent Registration ──────────────────────────────────────────────

  /**
   * Register a new agent in the AGP registry.
   */
  async registerAgent(input: RegisterAgentInput) {
    return this.transport.post("/agp/v1/agents/register", { json: input });
  }

  /**
   * Get an agent by ID.
   */
  async getAgent(agentId: string) {
    return this.transport.get(`/agp/v1/agents/${agentId}`);
  }

  /**
   * List agents, optionally filtered by organization.
   */
  async listAgents(organizationId?: string) {
    const params = organizationId
      ? `?organizationId=${organizationId}`
      : "";
    return this.transport.get(`/agp/v1/agents${params}`);
  }

  // ── Governance Actions ──────────────────────────────────────────────

  /**
   * Request a governed action on behalf of an agent.
   */
  async requestAction(input: ActionRequestInput) {
    return this.transport.post("/agp/v1/governance/action", {
      json: { ...input, protocolVersion: "agp/v1" },
    });
  }

  /**
   * Get governance events for an agent.
   */
  async getGovernanceEvents(agentId: string) {
    return this.transport.get(`/agp/v1/governance/events/${agentId}`);
  }

  // ── Deployments ─────────────────────────────────────────────────────

  /**
   * Create a new deployment for an agent.
   */
  async createDeployment(input: CreateDeploymentInput) {
    return this.transport.post("/agp/v1/deployments", { json: input });
  }

  /**
   * List deployments for an agent.
   */
  async listDeployments(agentId: string) {
    return this.transport.get(`/agp/v1/deployments/${agentId}`);
  }

  // ── Fleets ──────────────────────────────────────────────────────────

  /**
   * Create a new fleet.
   */
  async createFleet(input: CreateFleetInput) {
    return this.transport.post("/agp/v1/fleets", { json: input });
  }

  /**
   * List fleets, optionally filtered by organization.
   */
  async listFleets(organizationId?: string) {
    const params = organizationId
      ? `?organizationId=${organizationId}`
      : "";
    return this.transport.get(`/agp/v1/fleets${params}`);
  }

  // ── Activity ────────────────────────────────────────────────────────

  /**
   * Get the activity stream for an agent.
   */
  async getActivity(agentId: string) {
    return this.transport.get(`/agp/v1/activity/${agentId}`);
  }
}
